pyhms/core/initializers.py

```diff
@@ -0,0 +1,127 @@
+from abc import ABC, abstractmethod
+from typing import Callable
+
+import numpy as np
+from pyhms.core.individual import Individual
+from pyhms.core.problem import Problem
+from pyhms.utils.samplers import sample_normal, sample_uniform
+from scipy.stats.qmc import LatinHypercube, Sobol
+
+
+def linear_scale(sample: np.ndarray, bounds: np.ndarray) -> np.ndarray:
+    return bounds[:, 0] + sample * (bounds[:, 1] - bounds[:, 0])
+
+
+class PopInitializer(ABC):
+    def __init__(self, bounds: np.ndarray | None = None):
+        self._bounds = bounds
+        self._sampler: Callable = None
```

Comment:
If I want to implement a completely new Initializer I have to do something like this:
```
class CustomInitializer(PopInitializer):
  def prepare_sampler(self, context: dict = None) -> None:
    self._sampler = my_custom_sampler
```
The problem is that I have no clue what's inside the `context`. I have to overwrite a private field of this class (it doesn't sound good). I also have to read the source code to understand that my sampler is a function called with a single argument `pop_size: int`.

Reply:
Here you use `._sampler` but in `UniformGlobalInitializer` `.sampler` is used.

Reply:
You also overwrite `sample_pop` (e.g. in `UniformGlobalInitializer`) which is not annotated as an abstract method.

Reply:
The functionality is very simple, but you need to read many lines of code to understand what's going on. I think that it's overengineered. I don't like this approach.

Reply:
Maybe a solution like this:
```
class PopulationInitializer:
    def __init__(self, sampler: Callable[[int], np.ndarray]):
        self.sampler = sampler

    def initialize(self, pop_size: int) -> np.ndarray:
        return self.sampler(pop_size)

def uniform_sampler(bounds: np.ndarray) -> Callable[[int], np.ndarray]:
    def sampler(pop_size: int) -> np.ndarray:
        return np.random.uniform(0, 1, (pop_size, len(bounds)))
    return sampler
```
would be simpler? we can generate multiple functions that create samplers (or a factory) and it's straightforward -- you have to just pass your function as an argument. The function takes one parameter `pop_size` and that's all.